kotlinx-coroutines-test/common/src/TestScope.kt

```diff
@@ -170,14 +208,56 @@ internal class TestScopeImpl(context: CoroutineContext) :
     private val uncaughtExceptions = mutableListOf<Throwable>()
     private val lock = SynchronizedObject()
 
+    override val backgroundScope: CoroutineScope =
+        CoroutineScope(coroutineContext + BackgroundWork + ReportingSupervisorJob {
+            if (it !is CancellationException) reportException(it)
+        })
+
+    /**
+     * Collect the uncaught exceptions where possible.
+     *
+     * The exceptions caught in a [CoroutineExceptionHandler] via [reportException] will be collected.
+     * Additionally, this will collect exceptions with which coroutines in [backgroundScope] failed,
+     * unless they already were reported via the [CoroutineExceptionHandler].
+     * This allows us to report the situations where a background coroutine silently failed. For example:
+     * ```
+     * @Test
+     * fun foo() = runTest {
+     *   backgroundWorkScope.async {
+     *     throw TestException()
+     *   }
+     *   backgroundWorkScope.produce<Unit> {
+     *     throw TestException()
+     *   }
+     *   delay(1)
+     * }
+     * ```
+     */
+    // Only call this in a `synchronized(lock)` block.
+    @Suppress("INVISIBLE_MEMBER")
+    private fun collectUncaughtExceptions(): List<Throwable> {
+        val exceptions = LinkedHashSet(uncaughtExceptions)
```

Comment:
In `JobSupport` we use an identity hash set to workaround throwables that throw exceptions from their `hashCode`/`equals`. 
I don't think it's worth fixing here, but it's definitely worth to file an issue here and I'll escalate it to MPP stdlib